From: Al Viro <viro@zeniv.linux.org.uk>
Date: Thu, 11 Sep 2014 18:55:50 -0400
Subject: [PATCH] move the call of __d_drop(anon) into
 __d_materialise_unique(dentry, anon)
Git-commit: 6f18493e541c690169c3b1479d47d95f624161cf
Patch-mainline: v3.17
References: bsc#984194

and lock the right list there

Cc: stable@vger.kernel.org
Signed-off-by: Al Viro <viro@zeniv.linux.org.uk>
Acked-by: NeilBrown <neilb@suse.com>

---
 fs/dcache.c |    7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

--- a/fs/dcache.c
+++ b/fs/dcache.c
@@ -2509,6 +2509,12 @@ static void __d_materialise_dentry(struc
 		INIT_LIST_HEAD(&dentry->d_u.d_child);
 
 	anon->d_parent = (dparent == dentry) ? anon : dparent;
+	if (likely(!d_unhashed(anon))) {
+		hlist_bl_lock(&anon->d_sb->s_anon);
+		__hlist_bl_del(&anon->d_hash);
+		anon->d_hash.pprev = NULL;
+		hlist_bl_unlock(&anon->d_sb->s_anon);
+	}
 	list_del(&anon->d_u.d_child);
 	if (!IS_ROOT(anon))
 		list_add(&anon->d_u.d_child, &anon->d_parent->d_subdirs);
@@ -2566,7 +2572,6 @@ struct dentry *d_materialise_unique(stru
 				 * could splice into our tree? */
 				__d_materialise_dentry(dentry, alias);
 				write_sequnlock(&rename_lock);
-				__d_drop(alias);
 				goto found;
 			} else {
 				/* Nope, but we must(!) avoid directory
